ci(aio): add initial implementation for aio-builds setup

gkalpak opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[x] CI related changes
[ ] Other... Please describe:
```


**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```


**Other information**:
This is an initial (incomplete) implementation of the source code for setting PR previews.

Completed features/tasks:
- [x] Set up VM instance.
- [x] Snapshots should be accessible via HTTP.
  _(Works if set up on a server :smiley:)
- [x] One snapshot should be created per PR commit built on CI (e.g. triggering a new CI build by pushing a new commit, should create a new snapshot (while still keeping the previous snapshot available)).
- [x] Snapshots should be created even if the build failed.
  _(As long as the `aio` CI job passes.)
- [x] It should be able to map sub-paths to subdomains: `somedomain.com/angular-builds/prNUMBER-SHA/` → `prNUMBER-SHA.somePossiblyOtherDomain.com/`
- [x] Prevent overwriting existing content.
- [x] Snapshots should be removed once the corresponding PR is closed.
  _(A cron job will run periodically and remove closed/non-existent PRs.)_
- [x] Acquire/configure domain name.
  _(ngbuilds.io)_
- [x] Tests (e.g. for the Node.js scripts, e2e-like tests).
- [x] Use Google Cloud Platform's Persistent Disk for storage.
- [x] Once the snapshot is created, a link to the preview should appear on the PR as a comment.
- [x] Add support for HTTPS.
- [x] Create/Set up tokens/passwords (on VM, GitHub).
- [x] Prevent abuse of the preview feature. 
  _(Currently, only members of the angular-core team are allowed to deploy. Using Travis [JWT addon][1] and data from GitHub to verify.)_

Pending features/tasks:
_I created #14922 to track trings that need to be fixed/improved._

[1]: https://docs.travis-ci.com/user/jwt

gkalpak committed:
build(aio): update .gitignore

gkalpak committed:
ci(aio): add initial implementation for aio-builds setup

petebacondarwin commented on angular.io/aio-builds-setup/NOTES.md:
```diff
@@ -0,0 +1,57 @@
+
+## Run image
+
+- `sudo doker run \
```
`doker`?

IgorMinar replied:
doker -> docker

gkalpak replied:
:grin:

petebacondarwin commented on .gitignore:
```diff
@@ -18,15 +18,12 @@ modules/.vscode
 /modules/rollup-test/dist/
-
-# angular.io
-/angular.io/node_modules/
```
do we not install node modules to angular.io any more?

gkalpak replied:
It turns out this rule was not necessary, because we already have `node_module/` above (which will ignore `node_modules` anywhere, not just the top level).

petebacondarwin commented on angular.io/aio-builds-setup/dockerbuild/aio-builds.nginx-conf:
```diff
@@ -0,0 +1,21 @@
+  listen 80 default_server;
+  listen [::]:80;
+  return 404;
+}
```
So we have nginx and a node HTTP server. Are both needed? Is nginx just providing a front end?

gkalpak replied:
Originally, I started with nginx for serving, but we needed to run some scripts at some point for the more complex rules (such as preventing overwriting existing builds, unpacking and moving files around, etc.). I assume we could use shell scripts or something, but I thought Node (being JS) would be easier to maintain.

I've left the initial nginx config here (until we decide whether we want it or not), but currently it is not used.

IgorMinar commented on angular.io/aio-builds-setup/NOTES.md:
```diff
@@ -0,0 +1,57 @@
+
+## Questions
+
+- Do we need to run docker without sudo?
```
I don't think that running docker commands locally with sudo is a problem. We should however make sure that the account we use to transfer data from travis to this server is not a privileged account.

do you suspect any issues with using sudo?

gkalpak replied:
> do you suspect any issues with using sudo?

Not something specific :smiley:

IgorMinar commented on angular.io/aio-builds-setup/NOTES.md:
```diff
@@ -0,0 +1,57 @@
+
+- Do we need to run docker without sudo?
+- Do we need a `HEALTHCHECK` in Dockerfile?
+- How do we handle SSL/certificates in container?
```
SSL/certificates for what purpose?

1/ File transfer from CI to the preview server
OR
2/ for https serving of the preview files?

For the #1 we can just generate a certificate when building the VM. for #2 -could use https://letsencrypt.org/ ?

gkalpak replied:
I was thinking about #2, but #1 is reasonable too. 👍 for letsencrypt.

IgorMinar commented on angular.io/aio-builds-setup/dockerbuild/Dockerfile:
```diff
@@ -0,0 +1,49 @@
+EXPOSE 80 443
+
+# Go!
+CMD node ./index.js
```
is this node thing used only temporarily until we properly setup nginx?

IgorMinar replied:
if so, can you make a comment note about that please?

IgorMinar commented on angular.io/aio-builds-setup/dockerbuild/server/src/router.js:
```diff
@@ -0,0 +1,102 @@
+// Exports
+const router = module.exports = express.Router();
+
+// Route: /create-build/<pr>/<sha>
```
why not just use `scp` to upload the builds into a write-only directory (similar to how Mac's drop box works)?

gkalpak replied:
Would work too. I just wanted to keep as much as possible into the Node server, until we finalize the design (e.g. we might go with uploading to Cloud Storage or something).

IgorMinar commented on angular.io/scripts/deploy-preview.sh:
```diff
@@ -0,0 +1,15 @@
+cd "`dirname $0`/.."
+yarn run build -- --prod
+tar --create --gzip --directory "$INPUT_DIR" --file "$OUTPUT_FILE" .
+curl -i -X POST --data-binary "@$OUTPUT_FILE" "$AIO_BUILDS_HOST/create-build/$TRAVIS_PULL_REQUEST/$TRAVIS_COMMIT"
```
the interesting thing about this setup is that we have more flexibility to prevent abuse. e.g. we could easily use JWT to secure the upload: https://docs.travis-ci.com/user/jwt

gkalpak replied:
Seems promising 🎉

IgorMinar reviewed (changes requested):
interesting. this is slightly different than what I had in mind, but it might actually be better.

can you please address the comments I left?

additionally, can you please look into JWT to see if it's something we could use.

lastly, we we do stick with our own http server for handling uploads, can you please setup some kind of service that will restart the node process should it go unresponsive? there is a bunch of solutions for that in npm.

IgorMinar commented:
Also, would it make sense to use some kind of shared storage that we could mount onto this machine? It's not critical but it could be better to keep the storage volume separate from the volume used by the machine running all the servers.

petebacondarwin commented on aio/scripts/deploy-preview.sh:
```diff
@@ -0,0 +1,15 @@
+
+INPUT_DIR=dist/
+OUTPUT_FILE=/tmp/snapshot.tar.gz
+# TODO(gkalpak): Update this with the actual URL.
```
Do you still need to do this?

gkalpak replied:
Not really (except for updating to `https`) :smiley:
Made the change and removed the comment.

petebacondarwin commented on aio/scripts/deploy-preview.sh:
```diff
@@ -0,0 +1,15 @@
+
+cd "`dirname $0`/.."
+yarn run build -- --prod
+tar --create --gzip --directory "$INPUT_DIR" --file "$OUTPUT_FILE" .
```
I think we ought to try to sign this file with a private key that we place on Travis so that we can then check that the file is both intact and has come from Travis when it arrives at the GCE.

gkalpak replied:
Not sure what we gain 😕

petebacondarwin replied:
This would prevent people from spoofing the push to the GCE, no?

gkalpak replied:
The problem is that people can do anything from a PR, so I don't think this would stop them.

petebacondarwin commented on aio/aio-builds-setup/docs/NOTES.md:
```diff
@@ -0,0 +1,28 @@
+
+
+## Run image
+- `sudo doker run \
```
is it really doker not docker?

petebacondarwin commented on aio/aio-builds-setup/docs/NOTES.md:
```diff
@@ -0,0 +1,28 @@
+- Currently, builds will only be remove when the PR is closed. It is possible to upload arbitrary builds (for non-existent commits) until then.
+- Instead of creating new comments for each commit, update the original comment?
+- Do we need a static IP address?
+- Do we care about persistent disk automatic backup?
```
I doubt we need backups since I guess we could recreate any previews that were lost?

gkalpak replied:
That is my thinking. It is not trivial to do (one needs to checkout and build all shas (for which there were Travis jobs) for all open PRs, but it is definitely doable (and probably not even necessary).
Just needed a confirmation that I didn't miss anything.

petebacondarwin commented on aio/aio-builds-setup/docs/NOTES.md:
```diff
@@ -0,0 +1,28 @@
+## Questions
+- Do we care to keep logs (e.g. cron, nginx, aio-upload-server, aio-clean-up, pm2) outside of the container?
+- Currently, builds will only be remove when the PR is closed. It is possible to upload arbitrary builds (for non-existent commits) until then.
+- Instead of creating new comments for each commit, update the original comment?
```
:+1:

petebacondarwin commented on aio/aio-builds-setup/docs/NOTES.md:
```diff
@@ -0,0 +1,28 @@
+# VM Setup Instructions
```
How about adding links to the specific docs for each of the items in the list below?

gkalpak replied:
Yeah, that is my plan. The docs is pretty much a work in progress (and at an early stage), so feel free to ignore it for now.

petebacondarwin commented on aio/aio-builds-setup/docs/01. VM setup - Set up docker.md:
```diff
@@ -0,0 +1,35 @@
+# VM Setup - Set up docker
+
+
+## Install docker
```
Are there two versions of the install below (Debian and Ubuntu) because GCE uses Debian and you were testing on Ubuntu?

gkalpak replied:
Exactly :grin:

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/Dockerfile:
```diff
@@ -0,0 +1,103 @@
+
+ENV AIO_BUILDS_DIR=/var/www/aio-builds             TEST_AIO_BUILDS_DIR=/tmp/aio-builds   \
+    AIO_GITHUB_TOKEN=                              TEST_AIO_GITHUB_TOKEN=                \
+    AIO_NGINX_HOSTNAME=example.dev                 TEST_AIO_NGINX_HOSTNAME=example.dev   \
```
Are these references to `example.dev` just boilerplate or were they supposed to be this way?
Alternative question: are these variables important and used?

gkalpak replied:
They are supposed to be like this (although `AIO_NGINX_HOSTNAME` could be changed to point to the actual hostname, i.e. `ngbuilds.io`). I'll explain why below.

Currently, `AIO_NGINX_HOSTNAME` is used to perform the container's health-check (esentially a script that runs periodically by docker and checks if everything is OK with the container) and `TEST_AIO_NGINX_HOSTNAME` is used in the `verify-setup.sh` script, which essentially is a suite of e2e tests that can be run from within a running container and check that everything behaves as expected throughout.

The `.dev` top-level domain currently points to localhost, but according to [this][1] it is probably not a good idea to rely on the current behavior. I will set up `dnsmasq` to explicitly resolve some domain to localhost, so it can be used internally.

##
Note tht any of these env vars can be overwritten when spinning up the container (e.g. `docker run --env "AIO_NGINX_HOSTNAME="clibuilds.io" ...`). The idea is that the same docker image can be easily reused by other projects (e.g. material, cli, etc). 

[1]: http://stackoverflow.com/questions/28199785/why-all-dev-domains-target-to-my-localhost

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/Dockerfile:
```diff
@@ -0,0 +1,103 @@
+RUN mkdir /var/log/aio
+
+
+# Add extra package sources
```
Not sure this is an accurate comment, since the lines below appear to be installing node and yarn?

petebacondarwin replied:
OK, I take it back, these lines do not install the packages but just make them available via `apt-get`, right?
In which case, where is node installed? I can see yarn is being installed below. Does that automatically install node too?

gkalpak replied:
Ehm...node is installed on the same command that installs yarn ([just a couple of lines above][1]).

[1]: https://github.com/angular/angular/pull/14330/files/14a976e93644650dd4bc60025ca88961fd7dcb04#diff-949c8c392439c189f281b1932b4240bcR44

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/package.json:
```diff
@@ -0,0 +1,38 @@
+  },
+  "dependencies": {
+    "express": "^4.14.1",
+    "jasmine": "^2.5.3",
```
shouldn't `jasmine` be a `devDependency`?

gkalpak replied:
:grin: Here's the thing:

I have implemented this "setup verification functionality" (available as `aio-verify-setup` from within the container) that runs some e2e-like tests and verifies that things works expected (e.g. that nginx listens to the expected addresses/ports and responds as expected, that it forwards requests to the node.js server, that files are created with the expected permissions and owners, that clean-up takes place as expected, stuff like that).

I haven't made up my mind whether this should be part of the start-up sequence (right now it is not - I have added and removed it a couple of times :smiley:), but if someone wants to be able to log into the container and run this tests, they need jasmine.

The options are:

1. Do `yarn install` (without the `--production` flag), so that all dependencies are always available inside the container.
   _Doesn't sound too tempting._
2. Move `jasmine` to the `depenencies`, so that it is always possible to run `aio-verify-setup` without a hassle.
   _That's the current situation._
3. Move `jasmine` to the `devDependencies` and require that someone has to run `yarn install` (installing the `devDependencies`) before being able to run `aio-verify-setup`.
   _Can definitely leave with that, as long as we agree that `aio-verify-setup` doesn't have to be part of the start-up sequence._

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/clean-up/build-cleaner.ts:
```diff
@@ -0,0 +1,91 @@
+import {GithubPullRequests} from '../common/github-pull-requests';
+
+// Interfaces - Types
+type BuildCleanerOptions = {
```
Any reason you used a `type` here? Interfaces are usually preferred.

gkalpak replied:
There is a reason, but it's not a good one:
`tslint` wants me to prefix all interfaces with an `I`. Using a type keeps `tslint` happy without having to use an `I` or having to tweak that rule.

I am fine changing to interfaces.
OOC, why/when are interfaces preferred?

petebacondarwin replied:
Types are generally preferred when you are doing something clever like a union, and of course when the type is not an object. Interfaces are preferred for objects as they can be extended, etc. There is some discussion here: https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#3.10

I would suggest changing the tslint config in any case - I am not a fan of interfaces starting with `I`.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/clean-up/build-cleaner.ts:
```diff
@@ -0,0 +1,91 @@
+  protected repoSlug: string;
+
+  // Constructor
+  constructor({buildsDir, githubToken, repoSlug}: BuildCleanerOptions) {
```
Any reason you used an options object here? You would save lines (and an extra type) if you did:

```ts
constructor(protected buildsDir: string, protected repoSlug: string, protected githubToken?: string) {
```

since you would not need to declare the protected members separately, and there would be less need to do the runtime checking since the compiler would ensure that the first two variables were provided.

gkalpak replied:
Mainly for flexibility (especially in the initial dev phase, where things were more in flux). Also to avoid argument super long argument list in the future, but destructuring defeats that point, I guess :smiley:

I am fine changing it if you think it is preferable.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/utils.ts:
```diff
@@ -0,0 +1,11 @@
+export const getEnvVar = (name: string, isOptional = false): string => {
+  const value = process.env[name];
+
+  if (!isOptional && !value) {
```
what if value === 0?

gkalpak replied:
The values can only be string (unless someone does very weird hacks). From [the docs][1]:

> Assigning a property on `process.env` will implicitly convert the value to a string.

[1]: https://nodejs.org/api/process.html#process_process_env

petebacondarwin reviewed (changes requested):
I believe that we should move to a "pull and build" approach using a sandboxed docker for each build.

Unless @IgorMinar wakes up and says no, I think it would be good to create a PR to do that.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/clean-up/build-cleaner.ts:
```diff
@@ -4,31 +4,15 @@ import * as path from 'path';
-
   // Constructor
-  constructor({buildsDir, githubToken, repoSlug}: BuildCleanerOptions) {
+  constructor(protected buildsDir: string, protected repoSlug: string, protected githubToken?: string) {
```
See, much fewer LOC :-)

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/github-api.ts:
```diff
@@ -6,29 +6,26 @@ import * as https from 'https';
+  [key: string]: string | number;
+}
 
 type RequestParamsOrNull = RequestParams | null;
```
Do we actually need this type? It looks like the `RequestParams` type is only used in the `get`, `post` and `buildPath` methods, which already mark the parameter with a `?` which makes them `RequestParams|undefined`. Is there a case where we actually send in a `null` rather than `undefined`?

https://www.typescriptlang.org/play/#src=interface%20X%20%7B%20f%3A%20string%3B%20%7D%0D%0Afunction%20foo(x%3F%3A%20X)%20%7B%20%7D%0D%0Afoo(undefined)%3B%0D%0Afoo(null)%3B

gkalpak replied:
I want to be able to do `post('/some/url', null, {some: 'data'})` instead of having to do `post('/some/url', undefined /* or {} */, {some: 'data'})`. Because we are using `strictNullChecks` (which is not the case in the typescript playground I assume), we can't use `null` unless we explicitly add it to the type.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/github-api.ts:
```diff
@@ -6,29 +6,26 @@ import * as https from 'https';
+  constructor(protected repoSlug: string, githubToken?: string) {
     if (!repoSlug) {
       throw new Error('Missing required parameter \'repoSlug\'!');
     }
```
This check is no longer needed

petebacondarwin replied:
Or is it to catch the empty string?

gkalpak replied:
80% Safe-guarding againt an empty string.
20% Remnant from the initial implementation in JS 😛

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/upload-server-factory.ts:
```diff
@@ -8,24 +8,19 @@ import {UploadError} from './upload-error';
+  public create(buildsDir: string): http.Server {
     if (!buildsDir) {
       throw new Error('Missing or empty required parameter \'buildsDir\'!');
     }
```
This check is no longer needed

petebacondarwin replied:
Actually I accept that you might want to check against the empty string

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/helper.ts:
```diff
@@ -17,10 +17,11 @@ const TEST_AIO_UPLOAD_MAX_SIZE = +getEnvVar('TEST_AIO_UPLOAD_MAX_SIZE');
+export interface CmdResult { success: boolean; err: Error; stdout: string; stderr: string; }
+export interface FileSpecs { content?: string; size?: number; }
+export interface TestSuiteFactory { (scheme: string, port: number): void }
+export interface VerifyCmdResultFn { (result: CmdResult): void; }
```
After all my banging on about using interfaces over types, I think that one off function types are actually better represented as types :-( (sorry). When I see `interface` I immediately think of an object rather than a function, and I have to look hard to notice the function signature. I would use interfaces to represent functions when there are additional properties on the function.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/helper.ts:
```diff
@@ -36,6 +37,10 @@ class Helper {
   protected cleanUpFns: CleanUpFn[] = [];
+  protected portPerScheme: {[scheme: string]: number} = {
+    http: this.nginxPortHttp,
+    https: this.nginxPortHttps,
```
I feel it would be clearer to comment out this line rather than do the check in `runForAllSupportedSchemes` below (lines 113-116).

gkalpak replied:
I didn't want to comment it out, because I wanted to have the placeholder/pending test added in `runForAllSupportedSchemes()` (so that Jasmine shows the pending specs). I have now re-enabled the tests anyway (haven't yet pushed the changes).

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/clean-up/build-cleaner.spec.ts:
```diff
@@ -134,8 +126,7 @@ describe('BuildCleaner', () => {
     beforeEach(() => {
-      const fakeReaddir = (_: string, cb: typeof readdirCb) => readdirCb = cb;
-      fsReaddirSpy = spyOn(fs, 'readdir').and.callFake(fakeReaddir);
+      fsReaddirSpy = spyOn(fs, 'readdir').and.callFake((_: string, cb: typeof readdirCb) => readdirCb = cb);
```
oooh! I didn't know that you could do `param: typeof otherThing`!

gkalpak replied:
Yeah! Can be very handy 😄

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/clean-up/build-cleaner.spec.ts:
```diff
@@ -8,24 +8,18 @@ import {GithubPullRequests} from '../../lib/common/github-pull-requests';
     });
 
 
     it('should throw if \'repoSlag\' is empty', () => {
```
repoWhat?

gkalpak replied:
Just making sure you read carefully enough 😛

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/clean-up/build-cleaner.spec.ts:
```diff
@@ -40,14 +34,12 @@ describe('BuildCleaner', () => {
-        spyOn(cleaner as any, 'getOpenPrNumbers').and.callFake(() => {
-          return new Promise((resolve, reject) => openPrsDeferred = {resolve, reject});
-        });
+      cleanerGetExistingBuildNumbersSpy = spyOn(cleaner as any, 'getExistingBuildNumbers').and.callFake(() => {
```
why does `cleaner` need to be cast here? I thought that `spyOn` took any as its first param?

gkalpak replied:
`spyOn` has gotten smart. It knows what type it gets passed and does not allow accessing non-public members.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/helpers.ts:
```diff
@@ -1,5 +1,5 @@
 declare namespace jasmine {
-  export interface IDoneFn extends Function {
+  export interface DoneFn extends Function {
```
+1

petebacondarwin replied:
I think that this is a good use of a function

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/tslint.json:
```diff
@@ -3,8 +3,8 @@
     "arrow-parens": [true, "ban-single-arg-parens"],
+    "interface-name": [true, "never-prefix"],
     "max-classes-per-file": [true, 4],
-    "max-line-length": [true, 100],
```
works for me :-)

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/tslint.json:
```diff
@@ -3,8 +3,8 @@
   "rules": {
     "array-type": [true, "array"],
     "arrow-parens": [true, "ban-single-arg-parens"],
+    "interface-name": [true, "never-prefix"],
```
+100

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-sh/health-check.sh:
```diff
@@ -1,13 +1,19 @@
+
 # Check services
 services=(
   rsyslog
```
do you want to check `dnsmasq` here too?

gkalpak replied:
Attention to detail 💯 
I wanted to, but it exits with a non-zero code (it has different exit codes depending on some internal state (maybe whether you use it for DHCP). In our case, it returns 1 although the DNS part works. So, I didn't want to complicate things, since this is indirectly checked in the following section (making requests to several URLs).

Maybe I could add a request to an external URL (e.g. google.com) to verify that resolution works for them too.

petebacondarwin reviewed (approved):
A few minor comments. Nothing blocking merging this IMO. I'll leave it up to @gkalpak to choose what to fix up and then mark the PR as `merge`.

gkalpak committed:
refactor(aio): use types intead of interfaces for functions

gkalpak committed:
test(aio): fix typos (repoSlag --> repoSlug)

gkalpak committed:
feat(aio): check resolution of external URLs in HEALTHCHECK

gkalpak committed:
style(aio): correctly type tuple

gkalpak committed:
build(aio): update TypeScript (and other dependencies)

gkalpak committed:
feat(aio): add support for HTTPS (certificates provided by host - fallback to self-signed)

gkalpak commented:
Squashed previously reviewed commits and added a few more (addressing feedback and adding initial support for HTTPS).

Still pending:

- Finishing JWT implementation
- Rebasing on master

gkalpak committed:
refactor(aio): add `assertNotMissingOrEmpty()` helper

gkalpak committed:
feat(aio): make `githubToken` mandatory for `GithubApi`

gkalpak committed:
feat(aio): implement `GithubApi.getPaginated()`

gkalpak committed:
fix(aio): remove unnecessary `repoSlug` parameter from `GithubApi`

gkalpak committed:
feat(aio): implement `GithubTeams`

gkalpak committed:
feat(aio): implement `BuildVerifier`

gkalpak committed:
feat(aio): support passing secrets as files to the docker container

gkalpak committed:
feat(aio): verify uploaded builds based on JWT from Travis

gkalpak committed:
style(aio): change mock value for consistency

gkalpak committed:
build(aio): always compile before running tests

gkalpak committed:
fix(aio): do not hardcode the domain in preview link comments

gkalpak committed:
build(aio): upgrade to latest and pin major pm2 version

gkalpak committed:
ci(aio): test aio-builds-setup as part of the `aio` job

gkalpak committed:
build(aio): allow passing parameters to `docker build`

gkalpak committed:
build(aio): allow overwriting env vars at build time

gkalpak committed:
ci(aio): bringing in #14888 (because no patience)

gkalpak committed:
ci(aio): fail the build if preview deployment fails

mary-poppins commented:
The angular.io preview for 7da2b82 is available [here][1].

[1]: https://pr14330-7da2b829a60aae41baacf805f140043b04618dfd.ngbuilds.io/

IgorMinar commented:
@gkalpak we have our first preview!!!

the sha "7da2b82" doesn't seem to match anything in this PR which is weird, but other than that this really seems to be working! good job!

gkalpak commented:
:tada: :tada: :tada:

7da2b82 seems to be the merge commit that GitHub automatically creates to check if the PR is mergeable. I'll look into it.

gkalpak commented:
For reference, this is the issue about the Travis JWT bug: https://github.com/travis-ci/travis-ci/issues/7417

mary-poppins commented:
The angular.io preview for 7eb40ba is available [here][1].

[1]: https://pr14330-7eb40ba97b595d1b8dbc2c33b70763aab67fc97c.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 36c7153 is available [here][1].

[1]: https://pr14330-36c71534d62521b46e5ff665805b35b26c8f5b41.ngbuilds.io/

gkalpak commented:
I have created #14922 to track things that need to be fixed/improved. I expect more to be added as we test this out.

gkalpak committed:
fix(aio): ensure `NGBUILDS_IO_KEY` is not printed

Gaining access to another PR's JWT, would allow faking that PR's author wrt to
GitHub team membership verification for as long as the JWT is valid (currently
90 mins).

gkalpak committed:
feat(aio): make it easier to create a docker image (less variables to overwrite)

gkalpak committed:
test(aio): fix e2e tests

gkalpak committed:
refactor(aio): use more intuitive log filepath

gkalpak committed:
docs(aio): document arg for auto-restarting the docker container on boot

gkalpak commented:
It occurred to me that we should not print `NGBUILDS_IO_KEY` to the log. Please make sure I haven't missed anything in 49b4bd9.

mary-poppins commented:
The angular.io preview for 9b50b63 is available [here][1].

[1]: https://pr14330-9b50b637b74b74195b73a7bb6a133562c67b57d8.ngbuilds.io/

IgorMinar commented:
@gkalpak can you please rebase this? I just changed a bunch of our CI scripts which conflicts with this PR. sorry!!

IgorMinar commented:
@gkalpak to clarify: I moved all the ci scripts into `scripts/ci` directory

gkalpak commented:
>@gkalpak can you please rebase this?

I will :/

> I just changed a bunch of our CI scripts which conflicts with this PR.

It is not the first time (and I have a feeling it won't be the last) 😛

> to clarify: I moved all the ci scripts into scripts/ci directory

I wish that were the only change you made 😛 

##
I will give you th time to make any followup PRs and rebase tomorrow 😎

IgorMinar commented:
@gkalpak let's get this PR into a good enough shape so that we can merge it in tomorrow morning PT

mary-poppins commented:
The angular.io preview for 145f875 is available [here][1].

[1]: https://pr14330-145f8753124e96d7c729955930317c24a5a84902.ngbuilds.io/

gkalpak commented:
Rebased.

mary-poppins commented:
The angular.io preview for ee13970 is available [here][1].

[1]: https://pr14330-ee13970b910dba023adcc64e33562a105a23a0c3.ngbuilds.io/

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/Dockerfile:
```diff
@@ -0,0 +1,154 @@
+ARG      AIO_NGINX_PORT_HTTPS=443
+ARG TEST_AIO_NGINX_PORT_HTTPS=4433
+ARG      AIO_REPO_SLUG=angular/angular
+ARG TEST_AIO_REPO_SLUG=test-repo/test-slug
```
All these things called `TEST_...` are for running tests against this setup, right? Rather than something else? So this test repo slug doesn't have to match anything real, right?

gkalpak replied:
Right.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/Dockerfile:
```diff
@@ -0,0 +1,154 @@
+    AIO_DOMAIN_NAME=$AIO_DOMAIN_NAME                   TEST_AIO_DOMAIN_NAME=$TEST_AIO_DOMAIN_NAME                   \
+    AIO_GITHUB_ORGANIZATION=$AIO_GITHUB_ORGANIZATION   TEST_AIO_GITHUB_ORGANIZATION=$TEST_AIO_GITHUB_ORGANIZATION   \
+    AIO_GITHUB_TEAM_SLUGS=$AIO_GITHUB_TEAM_SLUGS       TEST_AIO_GITHUB_TEAM_SLUGS=$TEST_AIO_GITHUB_TEAM_SLUGS       \
+    AIO_LOCALCERTS_DIR=/etc/ssl/localcerts             TEST_AIO_LOCALCERTS_DIR=/etc/ssl/localcerts-test             \
```
Any reason why these are not variables too? Is it that they are not used elsewhere?

gkalpak replied:
Yes. They are only used inside the Dickerfile.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/nginx/aio-builds.conf:
```diff
@@ -0,0 +1,66 @@
+  ssl_certificate_key {{$AIO_LOCALCERTS_DIR}}/{{$AIO_DOMAIN_NAME}}.key;
+
+  root             {{$AIO_BUILDS_DIR}}/$pr/$sha;
+  disable_symlinks on from=$document_root;
```
security feature?

gkalpak replied:
Exactly. Since we are allowing arbitrary uploads, I assume it is better not to allow symlinks. (This was from before trusting authors based on GitHub team membership, but even with trusted authors, I think it is better not to allow symlinks unless we have to.)

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/github-api.ts:
```diff
@@ -0,0 +1,110 @@
+    return `${pathname}${joiner}${search}`;
+  }
+
+  protected getPaginated<T>(pathname: string, baseParams: RequestParams = {}, currentPage: number = 0): Promise<T[]> {
```
Unless it has changed since I last looked, this this is not how the GH paging works.
There are actually response headers that provide the URL to the next page of data.
I think that this will be fine for now but it would be worth looking into a more accurate usage of the API, in case they change things.

gkalpak replied:
I took the idea for this implementation for somewhere (I think it was from https://github.com/angular/periscope, but I can't find it know). Anyway, I think you are right, there should be a bettr way to implement this using GitHub's headers (but this works).
I will list it on #14922.

gkalpak committed:
ci(aio): only deploy a preview if the PR touched relevant files

mary-poppins commented:
The angular.io preview for ba064d9 is available [here][1].

[1]: https://pr14330-ba064d9eea0731328096c95dafef0ec5f21547fa.ngbuilds.io/

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/package.json:
```diff
@@ -0,0 +1,43 @@
+    "~~test-only": "node dist/test",
+    "pretest": "yarn run build",
+    "test": "yarn run ~~test-only",
+    "pretest-watch": "yarn run build",
```
don't need the `run` bit in `yarn run ...` any more :-)

gkalpak replied:
I had a script named `clean` in another project and I found out (the hard way) that omitting `run` is only possible when your script isn't named the same as a build-in yarn command (I didn't know about `yarn clean`).
It probably won't make a difference in our case here, but was being cautious.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-sh/upload-server-prod.sh:
```diff
@@ -0,0 +1,15 @@
+
+# Start the upload-server instance
+# TODO(gkalpak): Ideally, the upload server should be run as a non-privileged user.
+#                (Currently, there doesn't seem to be a straight forward way.)
```
Can you update the `sudoers` file https://help.ubuntu.com/community/Sudoers with a specific entry for running this script? See http://stackoverflow.com/a/6905797/287070

petebacondarwin replied:
Or is it that `pm2` controls the user?

gkalpak replied:
`pm2` does not support running the specified node process as a different user. It runs as the same user as `pm2` (see https://github.com/Unitech/pm2/issues/2232).
And running the `pm2` service as a non-privileged user is not straight forward (broken?).

petebacondarwin commented on aio/scripts/deploy-preview.sh:
```diff
@@ -0,0 +1,32 @@
+exec 3>&1
+httpCode=$(
+  curl --include --location --request POST --silent --write-out "\nHTTP_CODE: %{http_code}\n" \
+       --header "Authorization: Token $NGBUILDS_IO_KEY" --data-binary "@$OUTPUT_FILE" "$UPLOAD_URL" \
```
is this the bit where we have to be careful not to publish the Token onto the Travis logs?

gkalpak replied:
Yes, this is one of the places I know of. (Another one is in `env.sh`.)

mary-poppins commented:
The angular.io preview for b06df3d is available [here][1].

[1]: https://pr14330-b06df3d05549664aafecbed1d36024e1b940693e.ngbuilds.io/

petebacondarwin commented on scripts/ci/env.sh:
```diff
@@ -72,21 +72,29 @@ if [[ ${TRAVIS:-} ]]; then
-    # don't print the key
+    # - not using use setEnvVar so that we don't print the key
+    # - we overwrite the value set by Travis JWT addon here to work around travis-ci/travis-ci#7223 for NGBUILDS_IO_KEY
     export SAUCE_ACCESS_KEY=693ebc16208a-0b5b-1614-8d66-a2662f4e
```
what key is not being printed here?
if it is the key that is in the code here, then there is little benefit in not printing it on Travis, given that it is published here on github??

gkalpak replied:
Yes, the `SAUCE_ACCESS_KEY` is not printed. Not sure why (since it is already available on the repo). BTW, I copied this from @IgorMinar's #14888.

petebacondarwin reviewed (approved):
I admit that I didn't read all the test specs but I can't see anything significantly flawed in this PR.
I think that we are still waiting on @gkalpak to implement some additional filtering about running deployments. But I would be happy to see this merged.

petebacondarwin reviewed (approved):
The last commit LGTM too.

gkalpak committed:
refactor(aio): move scripts into `scripts/` directory

gkalpak committed:
ci(aio): do not deploy PR if preconditions not met

This avoids incorrectly failing the build if the PR author is not a member of one of the whitelisted GitHub teams.

gkalpak commented:
I have added 3 new commits (the last one not really belonging here :confused:) for skipping uploading the artifacts if the PR author is not a member of  whitelisted team.

Should I squash everything touching `aio-builds-setup/` into one commit?

IgorMinar commented on scripts/ci/env.sh:
```diff
@@ -72,21 +72,32 @@ if [[ ${TRAVIS:-} ]]; then
+  export NGBUILDS_IO_KEY=${NGBUILDS_IO_KEY:-$SAUCE_ACCESS_KEY}
+
+  # Personal token generated by mary-poppins, with only `read_org` permission
+  export GITHUB_TEAM_MEMBERSHIP_CHECK_KEY=b4cfbf1e6061c1bce934f1e7c3c8463f5cd6e0a9
```
try using `a3736b4b9948fe7e852d54dbbc360f999b6373a6` instead

IgorMinar commented:
@gkalpak I updated the key and force-pushed to your branch.

mary-poppins commented:
The angular.io preview for 9af6f6c is available [here][1].

[1]: https://pr14330-9af6f6c4389c6fc213cbbf46507444c5c83f3dfc.ngbuilds.io/

gkalpak commented:
@IgorMinar, sorry, I think have accidentally reverted one of you changes (which is strange because I thought `--force-with-lease` would prevent that, but maybe not when you have amended a commit).
I have restored the correct key I think (waiting for Travis to confirm).

mary-poppins commented:
The angular.io preview for 5fe5fad is available [here][1].

[1]: https://pr14330-5fe5fad311ab22436ccfbf9750c9e8d7023f343a.ngbuilds.io/

gkalpak commented:
:sparkles: :joy: :sparkles:
